### ISSUE TITLE
Issues with dynamically created grad_fn for views

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#8874]
Currently the dynamically created `grad_fn` to point to `base`. When a view is created via 
```py
base = torch.randn(3, requires_grad=True)
view1 = base[:2]
my_view = view1[:]
```
Then, if `base` is modified in-place, the `grad_fn` of `my_view` will be created to just point to `base`. Then the `retain_grad` and backward hooks on `view1` will be wrong.

E.g.,
```py
x = torch.ones(4, requires_grad=True).clone()
xx = x[2:]
y = xx[1]
xx.retain_grad()
#x[0] += 0
y.backward()
xx.grad
```

(Un)comment the line `x[0] += 0` changes the grad of `xx`.

This was previously reported at #8626 , but that also describes another unrelated (now solved) issue. So I'm opening a new one to track this.

**Comment by mruberry** [comment]
@albanD is this still an issue?

**Comment by soulitzer** [comment]
The line `x[0] += 0` should raise an error here. Something like - `RuntimeError: a view of a leaf Variable that requires grad is being used in an in-place operation.`

And it actually does if you remove `.clone()`.

**Comment by SsnL** [comment]
@soulitzer yeah I know. And it is totally unrelated with this issue. The point is about any view of a graph interior node, whether it is a clone or a deep net does not matter.

**Comment by albanD** [comment]
Actually, even though it is not obvious, I would argue that this is expected behavior :) 
In particular, the `xx` that you call `.retain_grad()` on is not the same as the one you try to access the grad of later (the inplace changes that).

You can see the example in this sample:
```python
# !pip install torchviz
from torchviz import make_dot
import torch

a = torch.rand(10, requires_grad=True).clone()

b = a.view_as(a)

out1 = b.exp()
a += 1
out2 = b.log()

make_dot(out1 + out2)
```

That generates the following graph
<img width="306" alt="image" src="https://user-images.githubusercontent.com/6359743/103353312-88674e00-4a76-11eb-815b-ef1b3b0f5e6d.png">

Here the out of place op on the left side shows that the "old" version of b (input to exp) is completely unrelated to the "new" one on the right side (input to log).
So calling retain_graph on the old one is expected not to do anything if only the new one is used. But if you use the "old" one in your code sample in an out of place op, then the retain_grad hook will actually be triggered and the .grad of the "old" variable will be populated.
